bndtools.core/src/bndtools/central/QueryWhiteboard.java

```diff
@@ -0,0 +1,54 @@
+package bndtools.central;
+
+import org.eclipse.search.ui.IQueryListener;
+import org.eclipse.search.ui.NewSearchUI;
+import org.eclipse.ui.IWorkbench;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+/**
+ * This component adds whiteboard facility to IQueryListeners.
+ */
+@Component(immediate = true)
+public class QueryWhiteboard implements ServiceTrackerCustomizer<IQueryListener, IQueryListener> {
+	@Reference
+	IWorkbench										wb;
+
+	ServiceTracker<IQueryListener, IQueryListener>	listeners;
+
+	BundleContext									bc;
+
+	@Activate
+	void activate(BundleContext bc) {
```

Comment:
```suggestion
	@Activate
	public QueryWhiteboard(BundleContext bc) {
```
will cause activation to occur before any bind methods are called.

Reply:
👍 thanks a bunch @rotty3000, I think that will do the trick! I'll try it out tomorrow.

Reply:
This is a solution. You will need to update the org.osgi.service.component.annotations dependency to 1.4.0 since that is where constructor injection support is added. The Eclipse version we support (2020-06) uses a version of Felix SCR which supports DS 1.4. So we can update all our Bndtools project dependencies on org.osgi.service.component.annotations to 1.4.0.

Reply:
to be much more clear, the entire solution which ensures workspace is present at activation before any bind methods are called looks like:
```java
final IWorkbench wb;
@Activate
public QueryWhiteboard(BundleContext bc, @Reference IWorkbench wb) {
 ...
 }
 ```
 as well as ensuring ds version 1.4.0.

Reply:
Done, thanks! I think I would probably make greater use of constructor injection in the future due to its ability to leverage the extra safety of `final` for the referenced services.